[airq] Air-Q binding Initial contribution

aurelio1 opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

I hereby add the new binding airq for use with the Air Analyzer and Logger air-Q

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

This is a new binding as my first contribution to the OpenHAB project.
I already tried a pull request a week ago and it was partly reviewed by @Hilbrand . After that, I accidentally retired the pull request. However, I considered all inputs @Hilbrand had entered up to then.

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work
-->

<!-- TESTING -->

It says that it is good practice to add an URL to the JAR, however I do not know where to publish it. It is not possible to publish it in this branch, otherwise it interferes with the pull request. So, where should I publish it?

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

Hilbrand commented:
To publish a jar you can create a release (see right pane on your openhab-addons repository page) and add the jar to the release.

aurelio1 commented:
> To publish a jar you can create a release (see right pane on your openhab-addons repository page) and add the jar to the release.

Thanks, worked.

Hilbrand commented:
Any reason you closed this pr?

aurelio1 commented:
I thought that I closed the comment...

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,229 @@
+
+## Discovery
+
+Auto-discovery is not possible in this version. Since the binding has to be configured at least with the password of the device, auto-discovery would be of limited value anyway.
```
It is fine to not implement discovery, however is it possible to add discovery and add the device without the password? It should then just go OFFLINE and have a status that the password is wrong/missing. The user then just edits the thing to have a password and it can then go ONLINE.

aurelio1 replied:
The implementation of the discovery on the air-Q is something odd, it does not answer to multi- or broadcasts. At this moment, I would like to publish the binding without the discovery and implement it later, once I get the appropriate information from the air-Q manufacturers. So at the moment it does not work without giving the IP address to the thing.

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,229 @@
+| Parameter              | Description                   |
+|------------------------|-------------------------------|
+| id                     | Device ID                     |
+| air-Q-Hardware-Version | Hardware version              |
+| air-Q-Software-Version | Firmware version              |
+| sensors                | Available sensors             |
+| SensorInfo             | Information about the sensors |
+| Industry               | Industry version              |
```
Parameters should be camelCase, so consider changing them to something like this:

SensorInfo to sensorInfo
Industry  to industryVersion
air-Q-Hardware-Version to hardwareVersion

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,229 @@
+
+| channel                   | type     | description                                                         |
+|---------------------------|----------|---------------------------------------------------------------------|
+| DeviceID                  | String   | Individual ID of the device                                         |
```
All channels should be camelCase. Consider changing to deviceID and Status to just `status`. etc...

Skinah reviewed (commented):
This will be a large amount of changes so will continue review after these are handled.

aurelio1 commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,229 @@
+
+## Supported Things
+
+One only Thing is supported: the air-Q device.
```
Why the `air-Q`?

Skinah replied:
You used air-Q as the label for the thing definition, so this is what will show up in the main UI when adding the device. Not sure I understand what you are asking here.

aurelio1 replied:
I mean the inverted comma, quote or however you call it, what surrounds air-Q in your suggestion. I can't write it in the text field because it starts a code section. You mean to use the quote around `air-Q` only when the device is meant, or whenever I use that name in the README.md?

Skinah replied:
I was using it like italics, underlining the word. But that is not necessary, and if you dont like the use of it, just leave the quote out.

aurelio1 commented:
Thanks @Skinah for the good suggestions - it was quite a work changing the names of the channels.
I am not able to fulfill the DCO requirement in the footer of the PR. I did the commands suggested by GitHub (`git commit --amend --no-edit --signoff` and `git push --force-with-lease origin 3.0.x`, but the error persists. What should I do?
Concerning the binding itself, I am ready for the next wave of review.

Skinah commented:
Yes it is a lot of work to change (I have done it myself and learnt the hard way) this is why:
1. The guidelines are here and that guideline is the very first on the list. https://docs.openhab.org/v2.1/developers/development/guidelines.html
2. I choose to wait until you had done them so I can look for anything missed when I continue looking the code over. Easy to miss it in the readme as the compiler wont complain :)

To fix the DCO signoff, the method I use is this one:
https://stackoverflow.com/questions/8981194/changing-git-commit-message-after-push-given-that-no-one-pulled-from-remote
Scroll down to the reply labelled as "To edit a commit other than the most recent:" Luckily the example they give is for 3 commits back and that happens to be where the issue is for you.

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,225 @@
+## Discovery
+
+Auto-discovery is not possible in this version.
+Since the binding has to be configured at least with the password of the device, auto-discovery would be of limited value anyway.
```
Suggest removing this sentence and just simply state "Auto-discovery is not yet supported".

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,225 @@
+
+| channel                   | type     | description                                                         |
+|---------------------------|----------|---------------------------------------------------------------------|
+| Status                    | String   | Status of the sensors                                               |
```
```suggestion
| status                    | String   | Status of the sensors                                               |
```

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,225 @@
+| dCO2dt                    | Number   | Change of CO<sub>2</sub> concentration                              |
+| dHdt                      | Number   | Change of Humidity                                                  |
+| dewpt                     | Number   | Dew Point                                                           |
+| door_event                | Switch   | Door Event (experimental)                                           |
```
```suggestion
| doorEvent                | Switch   | Door Event (experimental)                                           |
```

Skinah replied:
This has not been changed in the XML either.

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,225 @@
+| door_event                | Switch   | Door Event (experimental)                                           |
+| health                    | Number   | Health Index                                                        |
+| humidity                  | Number   | Humidity in percent                                                 |
+| humidity_abs              | Number   | Absolute Humidity                                                   |
```
humidityAbsolute or asbsoluteHumidity. This has not been changed in the XML either.

Skinah commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,225 @@
+| health                    | Number   | Health Index                                                        |
+| humidity                  | Number   | Humidity in percent                                                 |
+| humidity_abs              | Number   | Absolute Humidity                                                   |
+| measuretime               | Number   | Milliseconds needed for measurement                                 |
```
measureTime

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqBindingConstants.java:
```diff
@@ -0,0 +1,35 @@
+    // public static final ThingTypeUID THING_TYPE_SAMPLE = new ThingTypeUID(BINDING_ID, "sample");
+    public static final ThingTypeUID THING_TYPE_AIRQ = new ThingTypeUID(BINDING_ID, "airq");
+
+    // List of all Channel ids
+    public static final String CHANNEL_1 = "channel1";
```
Remove these unless you do have a channel1 ?... If you had used this example to define all the channel names, it would have been easier when you changed their name due to not using mixedCase naming. Most bindings do this.

aurelio1 replied:
I will do that in another binding I am developing, but I won't change it in this one. The issue is anyway changing the names in three files: airqHandler.java, README.md and thing-types.xml. The constant only helps for the .java file, afaik.

Skinah replied:
Yes only the java file, but if you have 10 java files and they are used in multiple places in each it can be a HUGE pain.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,956 @@
+                            "air-Q - airqHandler - handleCommand(): unknown command {} received (channelUID={}, value={})",
+                            command, channelUID, command);
+            }
+        } else if (command instanceof RefreshType) {
+            if (pollingJob != null) {
+                // pollingJob.notify();
+                // TODO: handle data refresh
+            }
+        }
```
Please remove all code that is commented out. Currently all these lines of code do nothing.

aurelio1 replied:
Concerning all commented code: some is not necessary and can be deleted. However, there are parts (at the end of the file) that are there because I am developing additional functions, but don't want to have them included in this version of the code because they don't work yet (getting historic files). How can I handle those code snippets?

Skinah replied:
I will leave that to the next person that does a review, as I do not know the answer.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,956 @@
+            }
+        }
+    }
+    // Getting the data files will be implemented in future, but it is not ready yet
+    /*
+     * private void getDataFiles() {
+     * Result res = null;
+     * String url = "http://".concat(ipaddress.concat("/dirbuff"));
```
The norm is to remove all commented out code as it will create a lot of lines that change when the comment block is removed.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,956 @@
+                String url = "http://".concat(ipaddress.concat("/config"));
+                String jsonbody = encrypt(jsoncmd.getBytes(), (String) (getThing().getConfiguration().get("password")));
+                String fullbody = "request=".concat(jsonbody);
+                // String testdecode = decrypt(jsonbody.getBytes(),
+                // (String) (getThing().getConfiguration().get("password")));
+                // logger.trace("air-Q - airqHandler - changeSettings(): testdecode={}, ", testdecode);
```
Remove all commented out code.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,956 @@
+                                dec.get(airqName).toString());
+                    }
+                    break;
+                // JsonArray calibarr = lastcalib.getAsJsonArray();
+                // logger.trace("air-Q - airqHandler - processType(): calibarr={}, isarr={}", lastcalib,
+                // lastcalib.isJsonArray());
+                // for (JsonElement el : calibarr) {
+                // logger.trace("air-Q - airqHandler - processType(): lastcalib element {}", el);
+
+                // }
```
Remove all commented out code.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,956 @@
+            @Override
+            public void run() {
+                logger.trace("air-Q - airqHandler - run(): starting polled data handler");
+                if ((!ipaddress.equals("")) && (!password.equals(""))) {
```
This I believe is more efficient as it does not compare two strings. Please change any other code to use this.

```suggestion
                if (!ipaddress.isEmpty() && !password.isEmpty()) {
```

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,956 @@
+            }
+        };
+
+        pollingJob = scheduler.scheduleWithFixedDelay(pollData, 0, POLLING_PERIOD_DATA, TimeUnit.MILLISECONDS);
```
```suggestion
        pollingJob = scheduler.scheduleWithFixedDelay(this::pollData, 0, POLLING_PERIOD_DATA, TimeUnit.MILLISECONDS);
```

Consider changing to the following:

Move the code that is in your runnable into a method like below. The "this::pollData" tells it to run the method so you don't need to define a runnable like you did.

public void pollData(){
//code here
}

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqConfiguration.java:
```diff
@@ -0,0 +1,28 @@
+
+public class airqConfiguration {
+
+    /**
+     * Sample configuration parameter. Replace with your own.
+     */
```
Remove the comments that the skeleton script created.

Skinah reviewed (commented):
Will go over it again once the channel names are fixed up.

aurelio1 commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,225 @@
+
+With the binding, it is possible to subscribe to all data delivered by the air-Q device.
+
+<img src="https://uploads-ssl.webflow.com/5bd9feee2fb42232fe1d0196/5e4a8dc0e322ca33891b51e4_air-Q%20frontal-p-800.png" alt="air-Q image" width="400px" height="324px" />
```
Yes, I have received permission for air-Q to use that link for the photo (by e-mail). I will ask again to be sure I can use a local copy of it. I put it in the resources folder.

aurelio1 replied:
Permission received from the CEO of air-Q to use a local copy of the image.

Skinah commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,694 @@
+		<state readOnly="true"></state>
+	</channel-type>
+
+	<channel-type id="WLAN_config_BSSID" advanced="false">
```
Consider making any channels that a lot of users will not want to link to an item as advanced="true"
The more channels that are linked, the more work the openHAB core has to do so it is an advantage to encourage users to not link everything.

aurelio1 replied:
I changed many more channels to 'advanced'. One question in this context: It would be great if there were a state `writeonly="true"`, but it seems OH does not have that. The air-Q has some parameters that should have a channel where a user can write to, but don't give back a value, e.g. the Log Level or the password. Is there a more elegant way to implement these than just having a channel providing the value NULL?

Skinah replied:
That would be best asked on the forum in the developer area as my knowledge is limited. I suspect the answer will be that you just tell the main UI to not display the state. There are also thing ACTIONS that you can implement into a binding which could match what your after in some cases perhaps?

https://www.openhab.org/docs/developer/bindings/#actions-bound-to-a-thing

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqBindingConstants.java:
```diff
@@ -0,0 +1,32 @@
+    private static final String BINDING_ID = "airq";
+
+    // List of all Thing Type UIDs
+    // public static final ThingTypeUID THING_TYPE_SAMPLE = new ThingTypeUID(BINDING_ID, "sample");
```
Remove this if not needed. No code should be commented out.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,936 @@
+                                channelUID.getId(), command.toString());
+                    }
+                    break;
+                // TODO
```
Is there is missing code here? remove comment if this is old.

aurelio1 replied:
Included now. I had the code in another, OH2 version, and forgotten to port it.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,936 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("air-Q - airqHandler - initialize(): ipaddress={}, password={}",
+                getThing().getConfiguration().get("ipAddress"), getThing().getConfiguration().get("password"));
```
Is there a need to place a users password into the logs?

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,936 @@
+                    if (decEl != null) {
+                        JsonObject decObj = decEl.getAsJsonObject();
+                        logger.debug("air-Q - airqHandler - run(): decObj={}", decObj);
+                        // processType(decObj, "bat", "battery", "pair");
```
remove commented code.

aurelio1 replied:
I added an explanation about why 'bat' is commented out. I have the same in thing-types.xml. The code ready to be used, but not applicable at this moment because no air-Q device is battery powered yet.

Skinah replied:
I'll leave this conversation open for the next reviewer to look at.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,986 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.info(
+                "air-Q - airqHandler - handleCommand(): request received to handle value {} command {} of channelUID={}",
+                command, command.getClass(), channelUID);
```
Info level logs should be used very very sparingly.
https://www.openhab.org/docs/developer/guidelines.html#f-logging
Also since the event.log already gets the commands and events, I don't believe it is needed to be logged again.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,986 @@
+                    // we do not allow to switch off Wifi because otherwise we can't connect to the air-Q device anymore
+                    break;
+                case "wifiInfo":
+                    newobj.addProperty("WifiInfo", ohCmd2airqCmd(command.toString()));
```
```suggestion
                    newobj.addProperty("WifiInfo", command == OnOffType.ON);
```
If the above suggestion works you eliminate 1. a conversion to a string and then 2. comparing two strings which is more costly then doing this. Same comment to all the times you use the ohCmd2airqCmd() function.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,986 @@
+                    break;
+                case "autoUpdate":
+                    // note that this property is binary but uses 1 and 0 instead of true and false
+                    newobj.addProperty("AutoUpdate", command.toString() == "ON" ? 1 : 0);
```
```suggestion
                    newobj.addProperty("AutoUpdate", command == OnOffType.ON ? 1 : 0);
```
No need to convert to a string.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,986 @@
+                            newobj.addProperty("WiFibssid", bssid);
+                        }
+                        newobj.addProperty("reset", wifidataobj.get("reset").getAsString());
+                        logger.info("air-Q - airqHandler - handleCommand(): dummy to change settings: {}",
+                                newobj.toString());
```
Should this be DEBUG or TRACE?

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,986 @@
+                                newobj.toString());
+                        changeSettings(newobj);
+                    } else {
+                        logger.error("Cannot extract wlan data from this string: {}", wifidatael);
```
ERROR should only be used when something very critical goes wrong. Link given to how to work out which logging level should be used, please review all uses of the logger and change them when you have some spare time as I wont comment on them all now.

aurelio1 replied:
I have reviewed all logging calls and replaced ALL .error with .warn as they are not critical. There is also no .info anymore, just .warn, .debug or .trace.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqConfiguration.java:
```diff
@@ -0,0 +1,27 @@
+ */
+
+@NonNullByDefault
+public class airqConfiguration {
```
Classes should always start with a capital so people know a class from its instance. Like this:

AirqConfiguration airqConfiguration = new AirqConfiguration();

You can instantly tell what is the class vs the instance.

aurelio1 replied:
OK. But eclipse requires to change the names of the files (also airqHandler ->AirqHandler), I hope that doesn't make your review unnecessarily complicated.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandler.java:
```diff
@@ -0,0 +1,986 @@
+        logger.debug("air-Q - airqHandler - initialize(): ipaddress={}",
+                getThing().getConfiguration().get("ipAddress"));
+        updateStatus(ThingStatus.UNKNOWN);
+        if (getThing().getConfiguration().get("ipAddress") != null) {
+            ipaddress = getThing().getConfiguration().get("ipAddress").toString();
+        }
+        if (getThing().getConfiguration().get("password") != null) {
+            password = getThing().getConfiguration().get("password").toString();
+        }
```
```suggestion
AirqConfiguration config = new AirqConfiguration(); //declare this globally at top of class.
//Lines below use here.
config = getConfig().as(AirqConfiguration.class);
            password = config.password; 
            ipaddress = config.ipaddress;
```

You don't need to use a local password and ipaddress variables as you can just use config.password and config.ipaddress instead anywhere you wish. Also it should be camelCase for ipAddress.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqConfiguration.java:
```diff
@@ -0,0 +1,27 @@
+
+@NonNullByDefault
+public class airqConfiguration {
+    public @NonNullByDefault({}) String ipaddress;
+    public @NonNullByDefault({}) String password;
```
```suggestion
    public String ipAddress="";
    public String password="";
```

Since you have it as ipAddress in the XML it needs to be identical here for the suggestion I made with "config" to work.

aurelio1 commented:
BTW, am I supposed to do 'resolve conversation' once I implement your suggestions, or do you do that when you check what I changed?

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,956 @@
+        config = getThing().getConfiguration().as(AirqConfiguration.class);
+        logger.debug("air-Q - airqHandler - initialize(): config={}", config);
+        updateStatus(ThingStatus.UNKNOWN);
+        if ((config.ipAddress.isEmpty()) || (config.password.isEmpty())) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "IP Address and the device password must be provided to access your air-Q.");
+            return;
```
This is unreachable code as in the XML file you marked these config parameters as being "required". The framework will not run this initialize() function until they hold a value and are not empty.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,956 @@
+            Arrays.fill(passkey, password.length(), 32, (byte) '0');
+        }
+        byte[] IV = Arrays.copyOf(encodedtextwithIV, 16);
+        // We comment this out because by setting the appropriate log level a third party could find the air-Q password.
+        // logger.trace("air-Q - airqHandler - decrypt(): passkey={}", passkey);
+        // logger.trace("air-Q - airqHandler - decrypt(): IV={}", IV);
+        // logger.trace("air-Q - airqHandler - decrypt(): text to decode: {}", ciphertext);
```
a general rule is not to merge commented out code, any code that is commented out should be removed, if it is needed to help people that may hit an issue then leave it as TRACE or DEBUG. Lastly just for knowledge if you have multiple logger calls like this in a row, you should test to see if TRACE is enabled in an if statement so the test is only done once and not three times.
Also agree to try and keep passwords out of the log, actually the same thought for base64 encoded material. It is very easy to reverse the encoding that it offers no protection at all in my opinion.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,956 @@
+                JsonObject jsonObj = ans.getAsJsonObject();
+                jsonAnswer = decrypt(jsonObj.get("content").getAsString().getBytes(), config.password);
+                if (jsonAnswer == null) {
+                    logger.warn("The air-Q data could not be decrypted. Probably the password is wrong.");
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Wrong password");
```
Can you do a find on 'updateStatus(' and if you have a logger and updateStatus( together can you work out if logger can be removed as it is redundant. See the developer guidelines for logging.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,956 @@
+        res = doNetwork(address, "GET", body);
+        if (res == null) {
+            if (thStatus != ThingStatus.OFFLINE) {
+                logger.warn("air-Q - airqHandler - run(): cannot reach air-Q device. Status set to OFFLINE.");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "air-Q device not reachable");
```
and here

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,956 @@
+                }
+            } catch (Exception e) {
+                if (System.out != null) {
+                    System.out.println(
+                            "air-Q - airqHandler - polldata.run(): Error while retrieving air-Q data: " + e.toString());
```
Use the logger or change the thing status and only one of those instead.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,956 @@
+                    }
+                }
+            } catch (Exception e) {
+                if (System.out != null) {
+
+                    System.out.println("Error in processConfigData(): " + e.toString());
+                }
```
and here

aurelio1 commented:
Applied all your suggestions.
I won't be able to do changes for the coming week. If you like to continue your review and place your comments, I will be glad and more than happy to implement them in a week.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,815 @@
+
+    public void pollData() {
+        logger.trace("air-Q - airqHandler - run(): starting polled data handler");
+        if ((!config.ipAddress.isEmpty()) && (!config.password.isEmpty())) {
```
config.ipAddress and config.password can not be empty as you have marked them as required in the XML file. The handler will not start until the framework is happy so you have this check multiple times and you can remove it which will free up CPU time. Use the find/search feature to find all cases.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,815 @@
+                case "time":
+                    String strstr = dec.get(airqName).toString();
+                    updateState(channelName, new StringType(strstr.substring(1, strstr.length() - 1)));
+                    logger.trace("air-Q - airqHandler - processType(): channel {} set to {}", channelName, strstr);
```
All of these TRACE logs are not needed and are double ups as all changes to the channels state is logged automatically in the event.log
Even if you don't have TRACE level turned on, this will still use up CPU time checking if it is on or off.
As mentioned it is done multiple times here so please look at them all under this 1 comment, thanks.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandlerFactory.java:
```diff
@@ -0,0 +1,70 @@
+ */
+@NonNullByDefault
+@Component(configurationPid = "binding.airq", service = ThingHandlerFactory.class)
+public class airqHandlerFactory extends BaseThingHandlerFactory {
```
Classes should start with a capital letter. AirqHandlerFactory please change any others to use a capital.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqBindingConstants.java:
```diff
@@ -0,0 +1,28 @@
+ * @author Aurelio Caliaro - Initial contribution
+ */
+@NonNullByDefault
+public class airqBindingConstants {
```
Capital A = AirqBindingConstants

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/airqHandlerFactory.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public @Nullable Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
+            @Nullable ThingUID thingUID, @Nullable ThingUID bridgeUID) {
+        logger.trace(
+                "air-Q - airqHandlerFactory - createThing: start with thingTypeUID={}, configuration={}, thingUID={}, bridgeUID={}",
+                thingTypeUID, configuration, thingUID, bridgeUID);
+        Thing th = super.createThing(thingTypeUID, configuration, thingUID, bridgeUID);
+        logger.trace("air-Q - airqHandlerFactory - createThing: result Thing={}", th);
```
Can these two logger lines be removed now the code is working?

aurelio1 replied:
I removed the second one, but kept the first. For the analysis of a system not recognizing a binding - in my experience one of the most common issues I had - it is crucial to see that the factory is called. Unfortunately the framework too often just does nothing when something is set wrong, instead of issuing a message. So I am happy to have this message when the factory is loaded. And it happens just once.

Skinah reviewed (changes requested):
OK I think I am getting close to all I can assist with once all the comments are addressed. I just have two more comments to make.

1. There is commented out code that you wish to leave in place as a WIP.
2. You still have some channels with non camelCase naming. Whilst some I believe should be changed, there are some that have an _ underscore in place of a decimal point which may be OK?!?!? to keep.

I suggest you look over the channel names again. 
Examples are 
WLAN_config_gateway
humidity_abs_maxerr
fineDustCnt02_5

Thanks for your contribution and hard work in getting this together.

aurelio1 commented:
@Skinah, thank you for the great suggestions! I learned a lot from your review and your annotations.
I have deleted all WIP code (and made a WIP copy for myself).
I also changed the channel names. However, I left the underscores of _maxerr or instead of the decimal point because a point is not allowed in a channel name, or at least that's what Eclipse thinks.

aurelio1 commented:
@Skinah  I hope I didn't do something wrong by re-requesting a review from you (before the second review by another reviewer). I just made a comment to your "change requested" and wanted to make sure you and the following reviewer would see that I included the changes you requested in your comment. Thank you again.

aurelio1 commented:
@Skinah May I ask you to review my changes one last time? I guess I implemented all your change requests, so you could forward this PR to the second reviewer without a change request. Thank you.

Skinah commented:
Sorry just busy and there are a lot of bindings all needing a review... Its normal to have to wait weeks for things to progress so don't stress. A second person can start a review at any time.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+        private final float value;
+        private final float maxdev;
+
+        public float getvalue() {
```
```suggestion
        public float getValue() {
```

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+            return value;
+        }
+
+        public float getmaxdev() {
```
```suggestion
        public float getMaxdev() {
```

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+            return maxdev;
+        }
+
+        // ResultPair() expects a string formed as this: [1234,56,789,012] and gives back a ResultPair
```
Can you make this a javadoc comment? then when you hover over this in an IDE it shows you this comment and helps others to understand the code quicker.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.trace(
```
All commands get logged in the event.log so this will be causing a higher load. Suggest this is removed.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+        pollingJob = scheduler.scheduleWithFixedDelay(this::pollData, 0, POLLING_PERIOD_DATA, TimeUnit.MILLISECONDS);
+        getConfigDataJob = scheduler.scheduleWithFixedDelay(this::getConfigData, 0, POLLING_PERIOD_CONFIG,
+                TimeUnit.MINUTES);
+        logger.debug("air-Q - airqHandler - initialize() finished");
```
This should be removed, see the dev guidelines on logging as it shows this as an example.

aurelio1 commented:
Sorry for my impatience!
I made the changes you suggested.

Skinah commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+                        subjson.addProperty("BrightnessNight", Float.parseFloat(command.toString()));
+                        newobj.add("NightMode", subjson);
+                        changeSettings(newobj);
+                    } catch (Exception exc) {
```
Its better to only catch the exact exception you are after.

Skinah reviewed (approved):
That's all I can help with.

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+# air-Q Binding
+
+The air-Q Binding integrates the air analyzer <a href="http://www.air-q.com">air-Q</a> device into the openHAB system.
```
Markdown syntax is preferred over HTML in this document, so that you can read it also unrendered. Same for the images.
```suggestion
The air-Q Binding integrates the air analyzer [air-Q](http://www.air-q.com) device into the openHAB system.
```

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+
+Auto-discovery is not yet supported.
+
+## Binding Configuration
+
+The binding does not need to be configured.
```
Then, you could remove this.

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+
+| channel                   | type     | description                                                         |
+|---------------------------|----------|---------------------------------------------------------------------|
+| status                    | String   | Status of the sensors                                               |
```
Can you mention the valid states?

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+This includes also the Maximum Error per sensor value.
+For the Maximum Error channels just add `_maxerr` to the channel names.
+
+| channel                   | type     | description                                                         |
```
You could make clear if a channel is read-only or writeable.

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+| channel                   | type     | description                                                         |
+|---------------------------|----------|---------------------------------------------------------------------|
+| status                    | String   | Status of the sensors                                               |
+| avgFineDustSize           | Number   | Average size of Fine Dust [experimental]                            |
```
You could use Units of Measure. For Example `Number:length`.

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+|---------------------------|----------|---------------------------------------------------------------------|
+| status                    | String   | Status of the sensors                                               |
+| avgFineDustSize           | Number   | Average size of Fine Dust [experimental]                            |
+| fineDustCnt00_3           | Number   | Fine Dust >0,3 &mu;m                                                |
```
As this is Unicode, you could use the actual character.
```suggestion
| fineDustCnt00_3           | Number:Length   | Fine Dust >0,3 µm                                                |
```

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+| fineDustCnt05             | Number   | Fine Dust >5 &mu;m                                                  |
+| fineDustCnt10             | Number   | Fine Dust >10 &mu;m                                                 |
+| co                        | Number   | CO concentration                                                    |
+| co2                       | Number   | CO<sub>2</sub> concentration                                        |
```
```suggestion
| co2                       | Number:Dimensionless   | CO<sub>2</sub> concentration                                        |
```

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+| co2                       | Number   | CO<sub>2</sub> concentration                                        |
+| dCO2dt                    | Number   | Change of CO<sub>2</sub> concentration                              |
+| dHdt                      | Number   | Change of Humidity                                                  |
+| dewpt                     | Number   | Dew Point                                                           |
```
And so on
```suggestion
| dewpt                     | Number:Temperature   | Dew Point                                                           |
```

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+| dCO2dt                    | Number   | Change of CO<sub>2</sub> concentration                              |
+| dHdt                      | Number   | Change of Humidity                                                  |
+| dewpt                     | Number   | Dew Point                                                           |
+| doorEvent                 | Switch   | Door Event (experimental)                                           |
```
Does the normal user know what is meant by this?

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+| wifiInfo                  | Switch   | Show WLAN status with LED                                           |
+| timeServer                | String   | Name of Timeserver address                                          |
+| location                  | Location | Location of air-Q device                                            |
+| nightmode_StartDay        | String   | Time to start day operation                                         |
```
Try to use a consistent naming scheme. Please check all.
```suggestion
| nightmodeStartDay        | String   | Time to start day operation                                         |
```

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+| nightmode_BrightnessNight | Number   | Brightness of LED at night                                          |
+| nightmode_FanNightOff     | Switch   | Switch off fan at night                                             |
+| nightmode_WifiNightOff    | Switch   | Switch off WLAN at night                                            |
+| devicename                | String   | Device Name                                                         |
```
```suggestion
| deviceName                | String   | Device Name                                                         |
```

fwolter commented on bundles/org.openhab.binding.airq/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,9 @@
+<features name="org.openhab.binding.airq-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-airq" description="airq Binding" version="${project.version}">
```
This is displayed when installing the binding.
```suggestion
	<feature name="openhab-binding-airq" description="air-q Binding" version="${project.version}">
```

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+    private @Nullable ScheduledFuture<?> pollingJob;
+    private @Nullable ScheduledFuture<?> getConfigDataJob;
+    private @Nullable ThingStatus thStatus;
+    protected static final int POLLING_PERIOD_DATA = 15000; // in milliseconds
```
It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+            JsonObject newobj = new JsonObject();
+            JsonObject subjson = new JsonObject();
+            switch (channelUID.getId()) {
+                case "wifi":
```
You could make these constants.

aurelio1 replied:
Do you think it is really necessary at this stage of the product? I will do it in another binding I am developing, from beginning, so i adds value to the development. I would prefer not to change them in this binding.

fwolter replied:
OK

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+        updateStatus(ThingStatus.UNKNOWN);
+        // We don't have to test if ipAddress and password have been set because we have defined them
+        // as being 'required' in thing-types.xml and OpenHAB will only initialize the handler if both are set.
+        String data = getDecryptedContentString("http://".concat(config.ipAddress.concat("/data")), "GET", null);
```
```suggestion
        String data = getDecryptedContentString("http://" + config.ipAddress + "/data"), "GET", null);
```

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+            IvParameterSpec ivSpec = new IvParameterSpec(IV);
+            cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);
+            byte[] decryptedText = cipher.doFinal(ciphertext);
+            content = new String(decryptedText);
```
Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+    public String encrypt(byte[] toencode, String password) {
+        String content = "";
+        logger.trace("air-Q - airqHandler - encrypt(): text to encode: {}", new String(toencode));
+        byte[] passkey = Arrays.copyOf(password.getBytes(), 32);
```
Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+        Result res = null;
+        res = doNetwork(address, "GET", body);
+        if (res == null) {
+            if (thStatus != ThingStatus.OFFLINE) {
```
Is there any reason for caching the status? You can retrieve it with `getThing().getStatus()`.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+        logger.debug("air-Q - airqHandler - doNetwork(): connecting to {} with method {} and body {}", address,
+                requestMethod, body);
+        try {
+            conn = (HttpURLConnection) new URL(address).openConnection();
```
You can save resources by using the framework's `HttpClient`. See https://github.com/openhab/openhab-addons/search?q=httpclientfactory for example.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+    public void pollData() {
+        logger.trace("air-Q - airqHandler - run(): starting polled data handler");
+        try {
+            String url = "http://".concat(config.ipAddress.concat("/data"));
```
See above. Replace `concat()`. Please check all.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+            String url = "http://".concat(config.ipAddress.concat("/data"));
+            String jsonAnswer = getDecryptedContentString(url, "GET", null);
+            if (jsonAnswer != null) {
+                Gson gson = new Gson();
```
You could store the Gson object to a final field to save resources.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+                    updateState(channelName, DateTimeType.valueOf(timestampString));
+                    break;
+                case "coord":
+                    JsonElement ans_coord = new Gson().fromJson(dec.get(airqName).toString(), JsonElement.class);
```
See above. Store Gson to a field.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+                    JsonElement jsonarr = new Gson().fromJson(dec.get(airqName).toString(), JsonElement.class);
+                    if ((jsonarr != null) && (jsonarr.isJsonArray())) {
+                        JsonArray arr = jsonarr.getAsJsonArray();
+                        String str = new String();
+                        for (JsonElement el : arr) {
+                            str = str.concat(el.getAsString()).concat(", ");
+                        }
```
When building Strings in loops, you could use `StringBuilder` to save resources.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,797 @@
+        Result res = null;
+        try {
+            String url = "http://".concat(config.ipAddress.concat("/config"));
+            String jsonbody = encrypt(jsoncmd.getBytes(), config.password);
```
Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandlerFactory.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    public @Nullable Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
+            @Nullable ThingUID thingUID, @Nullable ThingUID bridgeUID) {
+        logger.trace(
+                "air-Q - airqHandlerFactory - createThing: start with thingTypeUID={}, configuration={}, thingUID={}, bridgeUID={}",
+                thingTypeUID, configuration, thingUID, bridgeUID);
```
This is logged by the framework and can therefore be removed.

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>air-Q Binding</name>
+	<description>This is the binding for air-Q devices.</description>
```
Could you be a bit more verbose, so that somebody who hasn't heard about it gets a clue what this is about?

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>air-Q Binding</name>
+	<description>This is the binding for air-Q devices.</description>
+	<author>Aurelio Caliaro</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/i18n/airq_de_DE.properties:
```diff
@@ -0,0 +1,125 @@
+channel-type.airq.uptime.label = Laufzeit air-Q 
+channel-type.airq.uptime.description = Laufzeit air-Q
+
+channel-type.airq.bat_maxerr.label = Intervall Intervall Batteriestatus
```
```suggestion
channel-type.airq.bat_maxerr.label = Intervall Batteriestatus
```

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,695 @@
+		<category>Sensor</category>
+
+		<channels>
+			<!-- <channel id="DeviceID" typeId="devid"/> -->
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,695 @@
+			<channel id="logLevel" typeId="Logging"/>
+			<channel id="deleteKey" typeId="DeleteKey"/>
+			<channel id="fireAlarm" typeId="FireAlarm"/>
+			<!-- <channel id="air-Q-Hardware-Version" typeId="air-Q-Hardware-Version"/> -->
```
Can this be removed? Same for below commented out.

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,695 @@
+			</parameter>
+			<parameter name="password" type="text" required="true">
+				<context>password</context>
+				<label>Password of air-Q</label>
```
```suggestion
				<label>Password</label>
```

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,695 @@
+	</thing-type>
+
+
+	<!-- We have two places where the Device ID is delivered: with the measurement data and with the configuration.
+		We take the info from the configuration and show it as a property, so we don't need this at this moment. We
+		leave this as a reminder in case for some reason it will be needed in future, e.g. when an air-Q device
+		also sends data from other devices (then with another Device ID)
```
Please remove it so far. You could create an issue as reminder and/or stash this within Git for future releases.

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,695 @@
+
+	<channel-type id="typps" advanced="false">
+		<item-type>Number:Length</item-type>
+		<label>Average size of Fine Dust</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,695 @@
+		<state readOnly="true" pattern="%.3f %unit%"></state>
+	</channel-type>
+
+	<channel-type id="door" advanced="false">
+		<item-type>Number</item-type>
+		<label>Door Event (exp)</label>
+		<state readOnly="true"></state>
+	</channel-type>
```
Is this rather a trigger Channel? See https://www.openhab.org/docs/developer/bindings/thing-xml.html#trigger-channel-types

aurelio1 replied:
Thank you for pointing me to the trigger channels. This channel does not 'fire' an event, and I would prefer it to deliver this status about the door (if it works one day; currently it doesn't for me, my air-Q never signals an open door). Trigger channels don't have a status, so I would keep this as a regular channel.

fwolter replied:
OK

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,695 @@
+		<label>Calculate Maximum Errors</label>
+	</channel-type>
+
+	<!--
+		<channel-type id="getHistoryFiles" advanced="false">
+		<item-type>Switch</item-type>
+		<label>Retrieve History Files</label>
+		</channel-type>
+	-->
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,222 @@
+
+With the binding, it is possible to subscribe to all data delivered by the air-Q device.
+
+<img src="src/main/resources/image_air-Q.png" alt="air-Q image" width="400px" height="324px" />
```
Readme assets should be located in the `doc/` folder in the root of this binding.

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

Please add yourself to the `CODEOWNERS` file.
You need to add the binding to `bom\openhab-addons\pom.xml`.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

Please heed @Skinah's [comment ](https://github.com/openhab/openhab-addons/pull/10048#discussion_r593729165) about catching `Exception`. You should specify the exception you expect, otherwise all unchecked exceptions are cought, too. These are handled by the framework.

aurelio1 commented:
Thank you @fwolter for all those good suggestions. Once again I learned a lot. Using the HttpClient framework was challenging but I guess it helps for code efficiency.

fwolter commented:
You added only the image to the docs folder in your commit.

fwolter commented on CODEOWNERS:
```diff
@@ -237,6 +238,7 @@
 /bundles/org.openhab.binding.siemensrds/ @andrewfg
 /bundles/org.openhab.binding.silvercrestwifisocket/ @jmvaz
 /bundles/org.openhab.binding.sinope/ @chaton78
+/bundles/org.openhab.binding.siro2/ @Caliaro
```
Did you add this intentionally? Same for the pom.xml.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,772 @@
+                | InvalidAlgorithmParameterException | IllegalBlockSizeException exc) {
+            logger.warn("Error while decrypting. Probably the provided password is wrong.");
+            return null;
+        } /*
+           * catch (Exception e) {
+           * logger.warn("air-Q - airqHandler - decrypt(): Error while decrypting: {}", e.toString());
+           * return null;
+           * }
+           */
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandlerFactory.java:
```diff
@@ -0,0 +1,71 @@
+        return null;
+    }
+
+    @Override
+    public @Nullable Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
+            @Nullable ThingUID thingUID, @Nullable ThingUID bridgeUID) {
+        Thing th = super.createThing(thingTypeUID, configuration, thingUID, bridgeUID);
+        return th;
+    }
```
What's the purpose of this method?

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,660 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,660 @@
+	</channel-type>
+
+	<channel-type id="tvoc" advanced="false">
+		<item-type>Number</item-type>
```
```suggestion
		<item-type>Number:Dimensionless</item-type>
```

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,660 @@
+	</channel-type>
+
+	<channel-type id="uptime" advanced="true">
+		<item-type>Number</item-type>
```
```suggestion
		<item-type>Number:Time</item-type>
```

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,660 @@
+	<channel-type id="uptime" advanced="true">
+		<item-type>Number</item-type>
+		<label>Uptime</label>
+		<state readOnly="true" pattern="%.0f s"></state>
```
```suggestion
		<state readOnly="true" pattern="%.0f %unit%"></state>
```

aurelio1 replied:
This is not a time but a period, given in seconds. It seems that openHAB can handle time, but not periods (eg showing 120s as 2m). Or do you have another idea on how to handle this?
If openHAB can not handle time periods, I would prefer to have the unit displayed as seconds, otherwise the user does not know the unit.

fwolter replied:
`Number:Time` is a period. `DateTime` would be a point in time. If you prefer a special unit, you can leave the `s`.

aurelio1 replied:
Thank you. I put `Number:Time`, but left `s` as a unit because otherwise openHAB just shows the number, without any unit.

fwolter replied:
I would consider this a bug if no unit is displayed, although you specified `%unit%`. You could open an issue here https://github.com/openhab/openhab-core/issues

fwolter replied:
This isn't a bug in the framework. You need to send `QuantityType` instead of `DecimalType` if you want to use Units of Measure. See https://www.openhab.org/docs/concepts/units-of-measurement.html

aurelio1 replied:
OK. I tried the following:
xml file:
```
	<channel-type id="uptime" advanced="true">
		<item-type>Number:Time</item-type>
		<label>Uptime</label>
		<state readOnly="true" pattern="%d %unit%"></state>
	</channel-type>
```
code:
```
        updateState(channelName,
                new QuantityType<Time>(Integer.parseInt(dec.get(airqName).toString()), Units.SECOND));
```
However, it still does not work, the output it still just a number, without a unit. I have no experience in Parameterized Types (<>) so it would be the last thing assuming a bug - rather it is a deficiency from my side. I tried several variants to solve this, but I need your help. Where is the bug in above code? I would really love using this feature because abstracting a unit is a very good idea, but I'm stuck here.

aurelio1 replied:
Solved this, it is shown properly. The code above was correct, but I had not deleted the Item and created a new one. By recreating the item it works now.

fwolter commented on bundles/pom.xml:
```diff
@@ -5,563 +5,1109 @@
   <modelVersion>4.0.0</modelVersion>
 
   <parent>
+
```
Can you revert the formatting of this file?

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,224 @@
+
+## Supported Things
+
+Only one Thing is supported: The `air-Q` device.
```
Better use the Thing Type ID here.
```suggestion
Only one Thing is supported: The `airq` device.
```

aurelio1 replied:
Actually I tried to use airq when it was about the binding (because the binding name does not allow the hyphen), and air-Q when it is about the device. OK if I leave it like this?

fwolter replied:
I meant only this occurrence as users who use textual configuration need it, if that answers you question.

fwolter replied:
Can you comment on this?

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,224 @@
+
+## Thing Configuration
+
+The air-Q Thing must be configured with (both mandatory):
+| Parameter | Description                        |
```
There must be an empty line before and after tables.

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,224 @@
+img[src*="#small"] {
+    width:400px;
+    height:324px;
+}
```
Are you sure this will work? The GitHub preview renders this as oridnary text.

aurelio1 replied:
Right. I resized the image so it does not need a size modifier anymore.

fwolter commented:
Are you finished making all changes?

aurelio1 commented:
Yes. I think I have implemented all your suggestions, and even some more, as I applied QuantityType to more channels.

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,223 @@
+| fineDustCnt05             | Number:Dimensionless |    | Fine Dust >5 µm                                                     |
+| fineDustCnt10             | Number:Dimensionless |    | Fine Dust >10 µm                                                    |
+| co                        | Number:Dimensionless |    | CO concentration                                                    |
+| co2                       | Number               |    | CO₂ concentration                                                   |
```
Looks like you need to update some types here after making them `QuantityType`.

fwolter commented on bundles/org.openhab.binding.airq/README.md:
```diff
@@ -0,0 +1,223 @@
+| performance               | Number:Dimensionless |    | Performance Index (0 to 1000)                                       |
+| fineDustConc01            | Number               |    | Fine Dust concentration >1 µm                                       |
+| fineDustConc02_5          | Number               |    | Fine Dust concentration >2.5 µm                                     |
+| fineDustConc10            | Number               |    | Fine Dust concentration >10 µm             fni                      |
```
```suggestion
| fineDustConc10            | Number               |    | Fine Dust concentration >10 µm                                      |
```

aurelio1 replied:
Is this a typo?

fwolter replied:
I don't know. What is the purpose of the "fni"?

aurelio1 replied:
Sorry, I had understood that you inserted it. But actually it was _my_ typo, and I removed it now.

fwolter commented on bundles/org.openhab.binding.airq/src/main/java/org/openhab/binding/airq/internal/AirqHandler.java:
```diff
@@ -0,0 +1,793 @@
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import tec.uom.se.unit.Units;
```
This library shall not be used. You could use `org.openhab.core.library.unit.Units` instead.

fwolter commented on bundles/org.openhab.binding.airq/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,666 @@
+	</channel-type>
+
+	<channel-type id="mtime" advanced="true">
+		<item-type>Number</item-type>
```
```suggestion
		<item-type>Number:Time</item-type>
```

aurelio1 commented:
Just a brief comment to write that in my view all suggestions have been incorporated. I don't want to stress anyone, I just want to make sure there isn't a misunderstanding.

fwolter reviewed (approved):
LGTM

Now, you could add your binding's logo to the openHAB website. See https://deploy-preview-1540--openhab-docs-preview.netlify.app/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

aurelio1 commented:
Great, @fwolter , thanks a lot for your good suggestions and for your help. A new milestone for me ;)
